# $\alpha$=0.1

In [7]:
import torch
import torchvision.transforms as transforms              
from torchvision.datasets import CIFAR100      
from torch.utils.data import DataLoader
from src.inception_cifar100 import inceptionv3
from src.temperature_scaling import ModelWithTemperature
import numpy as np
from src.raps import raps_test


device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

dict_path = "C:\\Users\\jiayang\\ipynb\\trainedModel\\Inception_CIFAR100.pth"
model = inceptionv3()
model.load_state_dict(torch.load(dict_path, map_location=device, weights_only=True))
model.to(device)

# preprocess the images from CIFAR10
data_transform = transforms.Compose([
    transforms.ToTensor(),         
    transforms.Normalize((0.5071, 0.4865, 0.4409), (0.2673, 0.2564, 0.2762))  
])

# load data set from CIFAR100
dataset = CIFAR100(root="../../data", train=False, download=True,transform=data_transform)

# Temperature Scaling
temp_scal_loader = DataLoader(dataset, batch_size=32, shuffle=True)
model = ModelWithTemperature(model, temperature=0.5).to(device)
model.set_temperature(temp_scal_loader)
model.eval()

raps_test(model, dataset, device, num_runs=10, alpha=0.1, lambda_=0.02, k_reg=7)

Files already downloaded and verified
Before temperature - NLL: 1.387, ECE: 0.123
Optimal temperature: 0.633
After temperature - NLL: 1.877, ECE: 0.206
RAPS Classification, Start!

Running experiment 1/10...
q_hat = 0.9980654358863831
Total set size: 15655
Total coverage sets: 4540
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 3.131
Average Coverage Rate in runs 1: 0.908

Running experiment 2/10...
q_hat = 0.9977879881858825
Total set size: 15599
Total coverage sets: 4501
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 3.1198
Average Coverage Rate in runs 2: 0.9002

Running experiment 3/10...
q_hat = 0.9975217282772064
Total set size: 15368
Total coverage sets: 4489
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 3.0736
Average Coverage Rate in runs 3: 0.8978

Running experiment 4/10...
q_hat = 0.99721839427948
Total set size: 14792
Total coverage sets: 4473
Total samples amount: 5000
Average Predicti

# $\alpha$=0.3

In [7]:
raps_test(model, dataset, device, num_runs=10, alpha=0.3, lambda_=0.02, k_reg=5)

RAPS Classification, Start!

Running experiment 1/10...
q_hat = 0.8714932680130004
Total set size: 6397
Total coverage sets: 3484
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.2794
Average Coverage Rate in runs 1: 0.6968

Running experiment 2/10...
q_hat = 0.8679610311985014
Total set size: 6436
Total coverage sets: 3494
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.2872
Average Coverage Rate in runs 2: 0.6988

Running experiment 3/10...
q_hat = 0.8744889080524444
Total set size: 6417
Total coverage sets: 3484
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.2834
Average Coverage Rate in runs 3: 0.6968

Running experiment 4/10...
q_hat = 0.8685566127300262
Total set size: 6363
Total coverage sets: 3461
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 1.2726
Average Coverage Rate in runs 4: 0.6922

Running experiment 5/10...
q_hat = 0.8814487814903259
Total set size: 6

# $\alpha$=0.2

In [10]:
raps_test(model, dataset, device, num_runs=10, alpha=0.2, lambda_=0.02, k_reg=4)

RAPS Classification, Start!

Running experiment 1/10...
q_hat = 0.9650512099266053
Total set size: 9084
Total coverage sets: 4034
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.8168
Average Coverage Rate in runs 1: 0.8068

Running experiment 2/10...
q_hat = 0.9589962959289551
Total set size: 8896
Total coverage sets: 3970
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.7792
Average Coverage Rate in runs 2: 0.794

Running experiment 3/10...
q_hat = 0.9598138928413392
Total set size: 8910
Total coverage sets: 3986
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.782
Average Coverage Rate in runs 3: 0.7972

Running experiment 4/10...
q_hat = 0.9585777997970581
Total set size: 8770
Total coverage sets: 3964
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 1.754
Average Coverage Rate in runs 4: 0.7928

Running experiment 5/10...
q_hat = 0.964921748638153
Total set size: 9127


# $\alpha$=0.05

In [11]:
raps_test(model, dataset, device, num_runs=10, alpha=0.05, lambda_=0.01, k_reg=9)

RAPS Classification, Start!

Running experiment 1/10...
q_hat = 0.999998751282692
Total set size: 28494
Total coverage sets: 4774
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 5.6988
Average Coverage Rate in runs 1: 0.9548

Running experiment 2/10...
q_hat = 0.9999927312135697
Total set size: 25920
Total coverage sets: 4743
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 5.184
Average Coverage Rate in runs 2: 0.9486

Running experiment 3/10...
q_hat = 0.9999815464019776
Total set size: 24301
Total coverage sets: 4710
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 4.8602
Average Coverage Rate in runs 3: 0.942

Running experiment 4/10...
q_hat = 0.9999944239854812
Total set size: 26087
Total coverage sets: 4747
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 5.2174
Average Coverage Rate in runs 4: 0.9494

Running experiment 5/10...
q_hat = 0.9999954789876938
Total set size: 

#  Result
  
$\alpha$=0.1  
From the above test, following results can be collected :
- Final Average Prediction Set Size: 3.08
- Final Average Coverage: 90.04%  

$\alpha$=0.3  
From the above test, following results can be collected :
- Final Average Prediction Set Size: 1.28
- Final Average Coverage: 69.65%  

$\alpha$=0.2  
From the above test, following results can be collected :
- Final Average Prediction Set Size: 1.78
- Final Average Coverage: 79.71% 

$\alpha$=0.05  
From the above test, following results can be collected :
- Final Average Prediction Set Size: 5.47
- Final Average Coverage: 94.90%